code/contracts/ONEWallet.sol

```diff
@@ -351,25 +393,40 @@ contract ONEWallet is TokenTracker {
+    /// This function verifies that the first valid entry with respect to the given `eotp` in `commitLocker[hash]` matches the provided `paramsHash` and `verificationHash`. An entry is valid with respect to `eotp` iff `h3(entry.paramsHash . eotp)` equals `entry.verificationHash`
+    function _verifyReveal(bytes32 hash, uint32 indexWithNonce, bytes32 paramsHash, bytes32 eotp) view internal returns (uint32)
     {
-        Commit storage c = commitLocker[hash];
-        require(c.timestamp > 0, "Cannot find commit");
         uint32 index = indexWithNonce / maxOperationsPerInterval;
         uint8 nonce = uint8(indexWithNonce % maxOperationsPerInterval);
-        uint32 counter = c.timestamp / interval - t0;
-        require(counter == index, "Index - timestamp mismatch");
-        uint8 expectedNonce = nonces[counter];
-        require(nonce >= expectedNonce, "Nonce too low");
-        require(!c.completed, "Commit already completed");
-        require(c.paramsHash == paramsHash, "Invalid params hash");
-        // this should not happen (since old commit should be cleaned up already)
-        require(_isRevealTimely(c.timestamp), "Reveal too late");
+        Commit[] storage cc = commitLocker[hash];
+        require(cc.length > 0, "No commit found");
+        for (uint32 i = 0; i < cc.length; i++) {
+            Commit storage c = cc[i];
+            bytes32 expectedVerificationHash = keccak256(bytes.concat(c.paramsHash, eotp));
+            if (c.verificationHash != expectedVerificationHash) {
+                // Invalid entry. Ignore
+                continue;
+            }
+            require(c.paramsHash == paramsHash, "Parameter hash mismatch");
+            uint32 counter = c.timestamp / interval - t0;
+            require(counter == index, "Index - timestamp mismatch");
+            uint8 expectedNonce = nonces[counter];
+            require(nonce >= expectedNonce, "Nonce too low");
+            require(!c.completed, "Commit already completed");
+            // This normally should not happen, but when the network is congested (regardless of whether due to an attacker's malicious acts or not), the legitimate reveal may become untimely. This may happen before the old commit is cleaned up by another fresh commit. We enforce this restriction so that the attacker would not have a lot of time to reverse-engineer a single EOTP or leaf using an old commit.
+            require(_isRevealTimely(c.timestamp), "Reveal too late");
+            return i;
+        }
+        revert("No valid commit");
     }
 
-    function _completeReveal(bytes32 commitHash) internal {
-        Commit storage c = commitLocker[commitHash];
-        require(c.timestamp > 0, "Invalid commit hash");
+    function _completeReveal(bytes32 commitHash, uint32 commitIndex) internal {
+        Commit[] storage cc = commitLocker[commitHash];
+        require(cc.length > 0, "Invalid commit hash");
```

Comment:
This and the following line will never evaluate to false since the function verifyReveal ensures that returned index is always within the allowed range. The same for timestamp check. Since require is related to the workflow of the code and asserts check internal errors, I'd recommend here to use asserts instead of requires. So, it will not confuse the auditors seeking for the code logic.

Reply:
I agree